shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java

```diff
@@ -235,8 +270,15 @@ public void ensureInitializationCompleteAsync(
             new Runnable() {
               @Override
               public void run() {
-                if (resourceExtractor != null) {
-                  resourceExtractor.waitForCompletion();
+                InitResult result;
+                try {
+                  result = initResultFuture.get();
+                } catch (Exception e) {
+                  Log.e(TAG, "Flutter initialization failed.", e);
+                  throw new RuntimeException(e);
+                }
+                if (result.resourceExtractor != null) {
+                  result.resourceExtractor.waitForCompletion();
                 }
                 new Handler(Looper.getMainLooper())
                     .post(
```

Comment:
this is a comment for line 288. Is FlutterJNI.nativeInit the only line from ensureInitializationComplete that really needs to be run on the main thread? Could we potentially move everything into the init task to maximize the things not running on the main thread?

Reply:
The rest of `ensureInitializationComplete` is basically doing string manipulation to build the arguments passed to `nativeInit`.  The time consumed is negligible.